### DISCUSSION TITLE
Mock Client for Python SDK

### DISCUSSION THREAD

**Discussion by fel-este** [argoproj/argo-workflows#13507] [Q&A] [answered]
I saw there was a mock client for Go: https://github.com/argoproj/argo-workflows/discussions/7568
I couldn't find a mirroring object in the Python SDK. Is there one? The [test example](https://github.com/argoproj/argo-workflows/blob/ddbb3c7ad5b498d50514b3c1158ded56e333d75b/sdks/python/tests/client_unittest.py#L22) looks to be actually running workflows which I do not want.

**Comment by agilgur5** [comment, marked as answer]
The Python and Java SDKs are generated from the OpenAPI schema, so they don't have many features.
The Go SDK is used internally, so it's a bit different.

We also recommend the custom Hera Python SDK over the generated one: https://github.com/argoproj-labs/hera

**Reply by fel-este** [reply]
I do not see any mock argo client provided by Hera.

**Reply by agilgur5** [reply]
If one doesn't exist already, could certainly make a feature request in Hera's repo. Some similar topics seem to have been discussed before: https://github.com/argoproj-labs/hera/issues/277 (and even mocking steps themselves (vs networking) in https://github.com/argoproj-labs/hera/issues/971). 
The generated Python SDK is effectively deprecated in favor of Hera, as it's substantially better and supports much more than a generated tool can. 

The Go mock client is generated by a different, Go and k8s specific tool (upstream [`client-gen`](https://github.com/kubernetes/code-generator)), per the generated comments. It's also just a tiny wrapper around internal and k8s types (i.e. it uses source code). So we can't just re-use that for Python either.
